Probability sampling basics for telemetry events (#148)
open-telemetry/oteps: text/0148-sampling-probability.md

# Probability sampling of telemetry events

Specify a foundation for sampling techniques in OpenTelemetry.

## Motivation

In both tracing and metrics, there are widely known techniques for
sampling events that, when performed correctly, enable ways to lower
collection costs.  While sampling techniques vary, it is possible
specify high-level interoperability requirements that producers and
consumers of sampled data may follow to enable a wide range of
sampling designs.

## Explanation

Consider a hypothetical telemetry signal in which an API event
produces a unit of data that has one or more associated numbers.
Using the OpenTelemetry Metrics data model terminolgy, we have two
scenarios in which sampling is common.

1. Counter events: Each event represents a count, signifying the change in a sum
2. Histogram events: Each event represents an individual variable, signifying new membership in a distribution

> A Tracing Span event qualifies as both of these cases simultaneously.  It is

Author: I added an example using request bytes.

a Counter event (of 1 span) and a Histogram event (of 1 latency measurement).

In Metrics, [Statsd Counter and Histogram events meet this definition](https://github.com/statsd/statsd/blob/master/docs/metric_types.md#sampling).

In both cases, the goal in sampling is to estimate something about the
population of all events, using only the events that were chosen in
the sample.  Sampling theory defines various kinds of "estimator",
algorithms for calculating statistics about the population using just
the sample data.  For the broad class of telemetry sampling
application considered here, we need an estimator for the population
total represented by each individual event.

### Sample count adjustment

The estimated population total divided by the individual event value
equals the event's _adjusted count_.  The adjusted count is named

> `sample_count` where it appears as a field in a telemetry data type.

Reviewer: Why not `sample_weight`? Count (a) sounds like an integer, and (b) sounds like we literally received N identical events and compressed into one, while in practice we didn't even look at N-1 events and just assumed that this event represents N because of how it was selected.

Author: Considering all the feedback on this OTEP, I removed the recommendation of a specifically named field ("sample_count") in favor of a specific term used to describe the concept, "adjusted count". I reason that this is a count, not a weight, because it tells you how many times to count the event.

The field indicates that the individual sample event is estimated to

> represent `sample_count` number of identical events in the population.

Author: As for whether "count" implies an integer or not, I'm less sure. I get that idea, but I also think in terms of fractional counts quite easily. My odometer reads in 1/10ths of a kilometer. If we restrict this field to an integer, conceptually, we force the user to introduce bias or limit the kinds of sampling algorithm they can use. 

> Do we really intend to say that a positive sample_count value implies that there are similar-as-can-be duplicates that were elided, or just that this sample somehow represents a larger population?

Yes. Yes! It may sound like magic, but the statistical principles referenced here (in my not-a-real-mathematician-or-statistician kind of way) justify this interpretation.

I can see how a feeling of uneasiness creeps in. In your example, there were 35 events and we recorded only four of them. Suppose we had used simple 1-in-7 sampling but only 4 events were selected. The expected number of items in the sample was 5, but we only turned up 4. In this case we would estimate a total count of 28 for that period, which is off by 7. This is a form of variance: also known as sampling error.

There's a silver lining to this. When sampling is unbiased, sampling errors cancel out when estimating the mean. When we combine enough data, all the under-counting and over-counting averages out. For example, although in one period your 1-in-7 sampling scheme turns up 4 items, other periods will turn up 5 and 6 items, and over a long enough period of time, the sum of sample_count is expected to equal the sum of the population count. The same can be said when examining subsets of the data (see the [paper](https://dl.acm.org/doi/abs/10.1145/1314690.1314696) referenced above).

A standard sampling adjustment will be defined, and for it to work
there is one essential requirement.  The selection procedure must be
_unbiased_, a statistical term meaning that the process is expected to
give equal consideration to all possible outcomes.

The specified sampling adjustment sets the `sample_count` of each
sampled event to the inverse of the event's inclusion probability.
Conveying the inverse of the inclusion probability is convenient for
several reasons:

> - A `sample_count` of 1 indicates no sampling was performed
> - A `sample_count` of 0 indicates an unrepresentative event
> - Larger `sample_count` indicates greater representivity
> - Smaller `sample_count` indicates smaller representivity

Reviewer 2: ```suggestion
- A `sample_count` of 1 indicates no sampling was performed.
- A `sample_count` of 0 indicates an unrepresentative event.
- A larger `sample_count` indicates greater representativity.
- A smaller `sample_count` indicates smaller representativity.
```

- The sum of `sample_count` in a sample equals the expected 
  population size.
  
The zero `sample_count` value supports collecting values that were

> rejected from a sample to use as exemplars, which supports encoding of

Reviewer 2: Does "rejected from a sample to use as exemplars" mean that the event is _not_ exemplary, or that it's an example of something being elided? The list above says that a `sample_count` of zero indicates the former, using different language.

Author: I don't see these two (not exemplary and elided) as any different. The point I was trying to establish is that when sampling is in effect, and when the `sample_count` field is being interpreted from a stream of events, the value of zero is meaningful and necessary to ensure that the sum of sample_count equals the expected population count if we are to record a span that was not sampled.

A span with a count of zero should be considered a span that happened once but is not counted. You might use that span to complete another trace (rooted somewhere else), or you might use that span as an "auxiliary variable" (a term from Survey Analysis).

The specific use-case I have in mind when writing this stems from the weighted sampling algorithms mentioned here and above. If we want a sampling algorithm to compute a fixed size sample, when the number of events that will happen during an interval is unknown, there will be a chance that some item is first selected for a sample and later rejected from a sample (see [here](https://dl.acm.org/doi/10.1145/3147.3165)).

values that should not be counted in an estimate of the population
total.  This has applications to reservoir sampling designs, where
events may be selected for a sample only to be rejected before the end
of the frame.

### Sampling with attributes

Sampling is a powerful approach when used with event data that has
been annotated with key-value attributes.  It is possible to select

> arbitrary subsets of the sampled data and use it to estimate the count

Reviewer 2: ```suggestion
arbitrary subsets of the sampled data and use them to estimate the count
```
That is, assuming that "it" referred to the subsets. Alternately:
```suggestion
arbitrary subsets of the sampled data and use each to estimate the count
```

of arbitrary subsets of the population.

To summarize, there is a widely applicable procedure for sampling
telemetry data from a population:

- use an unbiased sampling algorithm to select telemetry events
- label each event in the sample with its `sample_count` (i.e., inverse inclusion probability)
- apply a predicate to events in the sample to select a subset
- apply an estimator to the subset to estimate the sub-population total.

Applied correctly, this approach provides accurate estimates for
population counts and distributions with support for ad-hoc queries
over the data.

### Changes proposed

This proposal leads to three change requests that will be carried out in
separate places in the OpenTelemetry specification.  These are:

> 1. For tracing, the SpanData message type should be extended with 

Reviewer 2: ```suggestion
1. _For tracing:_ The SpanData message type should be extended with 
```

   the `sample_count` field defined above.

> 2. For metrics aggregate data: Count information aggregated from

Reviewer 2: ```suggestion
2. _For metrics aggregate data:_ Count information aggregated from
```

   sample metric events will have floating point values in general.
   Histogram and Counter data must be able to support floating point
   values.

> 3. For metrics raw events: Exemplars should be extended with the 

Reviewer 2: ```suggestion
3. _For metrics raw events:_ Exemplars should be extended with the 
```

   `sample_count` field defined above.

### Example: Dapper tracing

Google's [Dapper](https://research.google/pubs/pub36356/) tracing
system describes the use of sampling to control the cost of trace
collection at scale.

### Example: Statsd metrics

> A Statsd counter event appears as a line of text, for example a metric

Reviewer 2: ```suggestion
A Statsd counter event appears as a line of text.

For example, a metric
```

named `name` is incremented by `increment` using a counter event (`c`)
with the given `sample_rate`.

```
name:increment|c|@sample_rate
```

For example, a count of 100 that was selected for a 1-in-10 simple
random sampling scheme will arrive as:

```
counter:100|c|@0.1
```

Probability 0.1 leads to a `sample_count` of 10.  Assuming the sample
was selected using an unbiased algorithm (as by a "fair coin"), we can
interpret this event as probabilistically equal to a count of `100/0.1
= 1000`.

## Internal details

The statistical foundation of this technique is known as the
Horvitz-Thompson estimator ("A Generalization of Sampling Without

> Replacement From a Finite Universe", JSTOR 1952).  The

Reviewer 2: ```suggestion
Replacement From a Finite Universe," JSTOR 1952).  The
```

Horvitz-Thompson technique works with _unequal probability sampling_
designs, enabling a variety of techniques for controlling properties
of the sample.  

For example, you can sample 100% of error events while sampling 1% of
non-error events, and the interpretation of `sample_count` will be
correct.

### Bias, variance, and sampling errors

There is a fundamental tradeoff between bias and variance in
statistics.  The use of unbiased sampling leads unavoidably to
increased variance.

Estimating sampling errors and variance is out of scope for this
proposal.  We are satisfied with the unbiased property, which guarantees
that the expected value of totals derived from the sample equals the
true population totals.  This means that statistics derived from
sample data in this way are always accurate, and that more data will
always improve precision.

### Non-probabilistic rate-limiters

Rate-limiting stages in a telemetry collection pipeline interfere with
sampling schemes when they operate in non-probabilistic ways.  When
implementing a non-probabilistic form of rate-limiting, processors

> MUST set the `sample_count` to zero.

Reviewer 2: It would help to present a scenario here where _not_ setting `sample_count` to zero predicts the wrong population value. It's not immediately obvious to the reader why rate limiting invalidates the estimate.

Author: Added a section on the use of zero adjusted count.

### No Sampler configured

When no Sampler is in place and all telemetry events pass to the
output, the `sample_count` field SHOULD be set to 1 to indicate

> perfect representivity, indicating that no sampling was performed.

Reviewer 2: ```suggestion
perfect representativity, indicating that no sampling was performed and that all events were preserved.
```

### Applicability for tracing

> The sampling approach covered here dictates how to select root spans

Author: I removed this section "applicability for tracing". I've made an example above for one way to use "adjusted count" or inverse inclusion probability in a tracing system, but this OTEP has exposed that there are many reasonable ways to do that. I believe another OTEP is needed specifically to address the use of sampling in a tracing system. This OTEP goes far enough to cover the metrics use-cases that I wanted covered.

in a probabilistic way.  When recording root spans, the `sample_count`
field should be set as described above.  The adjusted `sample_count`

> of the root span applies the trace, meaning the trace should be

Reviewer 2: ```suggestion
of the root span applies to the entire trace, meaning the trace should be
```

considered as representative of `sample_count` traces in the
population.

> When non-root spans are recorded because they are part of an ongoing

Reviewer 3: @jmacd It seems like most use cases discussed below support having non-root spans carry the same `sample_count` field as the trace. What do you think of just dropping the paragraph from this section?

Author: I dropped it!

trace, they are considered non-probabilistic exemplars.  Non-root

> spans should have `sample_count` set to zero.

Reviewer: This sounds like it's saying that the cumulative statistics (e.g. throughput or error rate) cannot be inferred for a service that is not a root?

Author: @akehlenbeck Agree! I left too much of a gap in this document by talking about what we record in the span event as opposed to what a tracing system can do to sample and re-sample spans at various levels in a trace. 

I remember the Dapper details you describe--unfortunately it wasn't discussed in the Dapper paper. I heard that @bogdandrutu may recall removing the `inverse_probability` field from trace propagation in the present-day Dapper system. (I hear other means are now used.)

The conversation about how to create new roots at interior nodes of a trace is important (and was to Dapper) when a high-throughput service occasionally calls a low-throughput service. Uniform sampling of the high-throughput service rarely traces the low-throughput service; starting new roots makes low-throughput interior nodes visible. 

In some ways, you could use all "true" roots in your trace collection to analyze the behavior of a certain class of interior span. This would be a lot like analyzing a stratified sample. You could also compute an independent sample for traces rooted at the same class of interior span.  You could use both data sets to analyze the behavior of that class of interior span, as they support estimatation of the same quantities.

Dapper chose a less expennsive alternative to what I wrote above,which allows spans sampled as parts of rooted traces to be mixed with spans sampled as parts of "new-roots".  I believe what Dapper did and you describe is probabilistically correct, and I believe the proposed `sample_count` field can be used to carry the intended meaning if this approach is taken.  The proof would be to show how the sume of `sample_count` over a stream of spans is expected to equal the true span count.

I think a new OTEP to discuss propagating "inclusion probability" (or its inverse) for recording in sample spans is called for.

Reviewer 4: @yurishkuro Reproducing the three possible cases in your example:
1. X and Y are sampled in 10% of all cases. The `sample_count` for Y would be 1.
2. X is not sampled while Y is sampled in 0.1 * 0.9 = 9% of all cases. The `sample_count` of Y would be 1 / (0.9 * 0.1).
3. X and Y are both not sampled in 81% of all cases. Y is not recorded.

The expectation of the sample count for Y would therefore be

10% * 1 + 9% * (1 / (0.9 * 0.1)) = 1.1

showing that there is something wrong. Unbiased sampling would give an expected sample count of 1.

This quite simple example already shows that it's not that easy to get subtree sampling right.

Reviewer 5: > I heard that @bogdandrutu may recall removing the `inverse_probability` field from trace propagation in the present-day Dapper system. (I hear other means are now used.)

Is anyone able to share details about these other means?

Reviewer 3: > So perhaps in case 1 the right thing to record is neither `p=0` like Josh wrote nor `p=1` like I wrote, but just `p=0.1 and Y isn't a root`, where the "Y isn't a root" part indicates that we're in case 1 and we shouldn't combine with other cases.

That makes a lot of sense to me and seems sufficient for tracing systems that don't support promoting child spans with missing parents. I'd also be curious if there's a way to combine this with case 2, but I don't have any ideas myself :-)

Author: This thread led me to remove a bunch of text about tracing specifics. I understand now that there are several ways to convey sampling information when collecting traces. I am now giving one example of how this can be used in a tracing system where only roots are sampled. I've also expanded the text about Dapper by a bit to explain this thread, somewhat.

### Applicability for metrics

The use of sampling in metrics makes it possible to record
high-cardinality metric events efficiently, as demonstrated by Statsd.

By pushing sampled metric events from client to server, instead of
timeseries, it is possible to defer decisions about cardinality
reduction to the server, without unreasonable cost to the client.

## Prior art and alternatives

The name `sample_count` is proposed because the resulting value is
effectively a count and may be used in place of the exact count.

Statsd conveys inclusion probability instead of `sample_count`, where

> it is often called "sample rate".

Reviewer 2: ```suggestion
it is often called "sample rate."
```

Another name for the proposed `sample_count` field is
`inverse_probability`, which is considered less suggestive of the

> field's purpose.

Reviewer 2: I find the latter to be clearer within the topic of sampling, though not fit for purpose here. I would prefer `sampling_factor` or `compression_factor`, because it's more general than "count"—in that it need not be integral—and it suggests multiplying by it to find another value.

Author: I've removed this name from my proposal. I'm using the term "adjusted count" to explain the concept, and would like you to review and see how this reads now.

# Probability sampling of telemetry events

Specify a foundation for sampling techniques in OpenTelemetry.

## Motivation

In tracing, metrics, and logs, there are widely known techniques for
sampling a stream of events that, when performed correctly, enable
collecting a tiny fraction of data while maintaining substantial
visibility into the whole population of events described by the data.

While sampling techniques vary, it is possible specify high-level
interoperability requirements that producers and consumers of sampled
data may follow to enable a wide range of sampling designs.

## Explanation

Consider a hypothetical telemetry signal in which an API event
produces a unit of data that has one or more associated numbers.
Using the OpenTelemetry Metrics data model terminology, we have two
scenarios in which sampling is common.

1. _Counter events:_ Each event represents a count, signifying the change in a sum.
2. _Histogram events:_ Each event represents an individual variable, signifying new membership in a distribution.

A Tracing Span event qualifies as both of these cases simultaneously.

> It is a Counter event (of 1 span) and at lesat one Histogram event

Reviewer 4: typo "lesat"

outcomes.

### Encoding inclusion probability

Some possibilities for encoding the inclusion probability, depending
on the circumstances and the protocol, briefly discussed:

1. Encode the adjusted count directly as a floating point or integer number in the range [0, +Inf).  This is a conceptually easy way to understand sampling because larger numbers mean greater representivity.
2. Encode the inclusion probability directly as a floating point number in the range [0, 1).  This is typical of the Statsd format, where each line includes an optional probability.  In this context, the probability is commonly referred to as a "sampling rate".  In this case, smaller numbers mean greater representivity.
3. Encode the negative of the base-2 logarithm of inclusion probability.  This restricts inclusion probabilities to powers of two and allows the use of small non-negative integers to encode power-of-two adjusted counts.
4. Fold the adjusted count into the data.  This is appropriate when the data itself carries counts, such as for OTLP Metrics Sum and Histogram points encoded using delta aggregation temporality.  This may lead to rounding errors, when adjusted counts are not integer valued.

> 

Reviewer 4: Maybe mention that the 4th possibility is different from the first 3 as you loose information about the variance. Without additional information, you no longer know if the count is exact or represents an estimate, and if the sampling rate was the same for all increments or not.

Author: Done.

events that took place whether the events were sampled or not, which
requires being able to recognize a zero value for the adjusted count
as being distinct from a raw event where no sampling took place.

#### Recognizing zero adjusted count

An adjusted count of zero indicates an event that was recorded, where
according to the sampling design its inclusion probability is zero.
These events are may be included in a stream of sampled events as
auxiliary information, and consu

Recording events with zero adjusted count are a useful way to record
auxiliary information while sampling, events that are considered
interesting but which are accounted for by the adjusted count of other
events in the same stream.

Consider a span sampling design that applies a sampling decision only
to the roots of a trace.  Non-root spans must be recorded when their
parent span is selected for a sample.  For a class of spans that
sometimes are and sometimes are not the root of a trace, we have three
outcomes:

1. Span is part of another trace: adjusted count is zero
2. Span is a root, was selected for the sample: adjusted count is non-zero
3. Span is a root, was not selected for the sample: not recorded.

> 

Reviewer 4: What I would expect, when aggregating the _adjusted counts_ of all spans of one particular service, is an unbiased estimate of the true number of requests to that service. Unfortunately this is not the case, if there are non-root spans for that particular service, because the sampling decision is made at root spans, and the corresponding sample rates and hence _adjusted counts_ are not known for non-root spans. Consequently, a zero _adjusted count_ would just indicate that the _adjusted count_ is not known and needs to be retrieved from the root span. However, it does not necessarily mean that the span does not contribute to the total count at all.

Therefore, I would prefer to propagate the sample rate that was used for some root span to its descendants, such that they know the sample rate and are able to set the _adjusted count_ properly.

Author: I was trying to illustrate an example, not to suggest a real tracing Sampler, however I can see that no one is interested in this case, so it's not a good example.

I removed these paragraphs. Instead, I added more depth to the description of the Dapper system, and how it propagated the parent's inclusion probability, particularly when it made a negative sampling decision, so the child could calculate the conditional probability needed to ensure an overall rate of its own traces. I sense there is more interested in this approach, now that tracing has reached GA.

of itself."  It relies on the use of an unbiased sampling design.
Readers are referred to [recommended reading](#recommended-reading)
for more resources on sampling with attributes.

### Summary: a general technique

Sampling is a powerful approach when used with event data that has
been annotated with key-value attributes.  It is possible to select
arbitrary subsets of the sampled data and use each to estimate the count
of arbitrary subsets of the population.

To summarize, there is a widely applicable procedure for sampling
telemetry data from a population:

- describe how to map telemetry events into discrete frames
- use an unbiased sampling design to select events
- encode the adjusted count or inclusion probability in the recorded events
- apply a predicate to events in the sample to select a subset of events
- sum the adjusted counts of the subset to estimate the sub-population total.

Applied correctly, this approach provides accurate estimates for
population counts and distributions with support for ad-hoc queries
over the data.

### Changes proposed

This OTEP proposes no formal changes in the OpenTelemetry

> specitication.  It is meant to lay a foundation for importing sampled

Reviewer 6: ```suggestion
specification.  It is meant to lay a foundation for importing sampled
```

# Probability sampling of telemetry events

Specify a foundation for sampling techniques in OpenTelemetry.

## Motivation

In tracing, metrics, and logs, there are widely known techniques for
sampling a stream of events that, when performed correctly, enable
collecting a fraction of the complete data while maintaining
substantial visibility into the whole population of events.

These techniques are all forms of approximate counting.  Estimates
calculated by the forms of sampling outlined here are considered
accurate, in that they are random variables with an expected value

> equal to the true count.  With enough observations, 

Reviewer 7: ```suggestion
equal to the true count.
```

This seems to be dangling here, is it missing somewhere down below?

Author: Removed. Thanks.

be controlled for by the sample size.  Variance is unavoidable in an
unbiased sample, but it vanishes when you have enough data.

Although this makes it sounds like small sample sizes are a problem,
due to expected high variance, this is just a limitation of the
technique.  When variance is high, use a larger sample size.

An easy approach for lowering variance is to aggregate sample frames
together across time.  For example, although the estimates drawn from
a one-minute sample may have high variance, combining an hour of
one-minute sample frames into an aggregate data set is guaranteed to
lower variance.  It must, because the data remains unbiased.

### Conveying the sampling probability

Some possibilities for encoding the adjusted count or inclusion
probability are discussed below, depending on the circumstances and
the protocol.

There are several ways of encoding this information:

- as a dedicated field in an OTLP protobuf message
- as a non-descriptive Attribute in an OTLP Span, Metric, or Log
- without any dedicated field.

#### Encoding adjusted count

We can encode the adjusted count directly as a floating point or

> integer number in the range [0, +Inf).  This is a conceptually easy

Author: A uniform sample of 40% has adjusted count of 2.5. To me this, makes sense, but I realize that it can present problems in some systems. This is mentioned in the example about computing Histogram metrics from spans, because OTLP Histogram points do not have a way to represent non-integer counts.  One of the reasons I like using an attribute to reflect adjusted count, not multiplying these values into the data itself, is that it allows the use of floating-point counts even where values are integers; OTOH, it punts the problem for the next consumer. I am happy to recommend the use of integer-reciprocal sampling rates in situations where this is a real problem.

> integer number in the range [0, +Inf).  This is a conceptually easy

Author: I added a paragraph to state the above. ^^^

samples are *random variables* that, when correct procedures are
followed, accurately reflect their true value, making them unbiased.
Unbiased samples can be used for after-the-fact analysis.  We can
answer questions such as "what fraction of events had property X?"
using the fraction of events in the sample that have property X.

This document outlines how producers and consumers of sample telemetry
data can convey estimates about the total count of telemetry events,
without conveying information about how the sample was computed, using
a quantity known as **adjusted count**.  In common language, a
"one-in-N" sampling scheme emits events with adjusted count equal to
N.  Adjusted count is the expected value of the number of events in
the population represented by an individual sample event.

## Examples

In all of these examples, the use of probability sampling leads to an
attribute named like `sampling.sampler_name.adjusted_count`.
Consumers of spans, metrics, and logs annotated with adjusted counts
are able to calculate accurate statistics about the population of
events, at a basic level, without knowing details about the sampling
configuration.

### Span sampling

<details> 

> <summary> Example use-cases for probability sampling of spans

Reviewer 8: ```suggestion
<summary>Example use-cases for probability sampling of spans
```

generally involve generating metrics from spans.</summary>

#### Sample spans to Counter Metric

For every span it receives, the example processor will synthesize
metric data as though a Counter instrument named `S.count` for span

> named `S` had been incremented once per span at the original `Start()`

Reviewer 8: One potential challenge here is that the span `Start()` will not have access to all the attributes (e.g. some of the attributes might be added later).

Author: I see now the consequences of moving the examples above the introductory text, some missing bits. I've added "complete" here to avoid the ambiguity. I'm trying to stick the OTel tracing data model, where spans encoded in OTLP are already finished.

call site.

This processor will add the adjusted count of each span to the
instrument (e.g., `Add(adjusted_count, labels...)`)  for every span it
receives, logically taking place at the start or end time of the span.

#### Sample spans to Histogram Metric

For every span it receives, the example processor will synthesize
metric data as though a Histogram instrument named `S.duration` for
span named `S` had been observed once per span at the original `End()`
call site.

The OpenTelemetry Metric data model does not support histogram buckets
with non-integer counts, which forces the use of integer adjusted
counts here (i.e., 1-in-N sampling rates where N is an integer).

> Logically spaking, this processor will observe the span's duration its

Reviewer 9: Typo

Reviewer 8: ```suggestion
Logically speaking, this processor will observe the span's duration its
```

adjusted count number of times for every span it receives, at the end
time of the span.

#### Sample span rate limiting

A collector processor will introduce a slight delay in order to ensure
it has received a complete frame of data, during which time it

> maintains a fixed-size buffer of input spans.  If the number of spans

Reviewer 8: Will this buffer only contain finished spans, or it can have spans that are started but not finished?

Author: I added "finished" to describe these spans.

the selected sample spans are output with possibly updated adjusted
counts.
</details>

### Metric sampling

<details> 
<summary> Example use-cases for probability sampling of metrics
are aimed at lowering cost and addressing high cardinality.</summary>

#### Statsd Counter

A Statsd counter event appears as a line of text, describing a
number-valued event with optional attributes and inclusion probability
("sample rate").

For example, a metric named `name` is incremented by `increment` using
a counter event (`c`) with the given `sample_rate`.

```
name:increment|c|@sample_rate
```

For example, a count of 100 that was selected for a 1-in-10 simple
random sampling scheme will arrive as:

```
counter:100|c|@0.1
```

Events in the example have with 0.1 inclusion probability have
adjusted count of 10.  Assuming the sample was selected using an
unbiased algorithm, we can interpret this event as having an expected
count of `100/0.1 = 1000`.

#### Metric exemplars with adjusted counts

The OTLP protocol for metrics includes a repeated exemplars field in
every data point.  This is a place where histogram implementations are

> able to provide example context to corrlate metrics with traces.

Reviewer 8: ```suggestion
able to provide example context to correlate metrics with traces.
```

single metric name, allowing no more than K distinct label sets per
export interval.  The export interval is fixed to a short interval so
that a complete set of distinct labels can be stored temporarily.

Caveats: as presented, this works for Sum and Histogram points
received with Delta aggregation temporality and where the Sum is
monotonic (see
[opentelemetry-proto/issues/303](https://github.com/open-telemetry/opentelemetry-proto/issues/303)).

Considering data points received during the interval, when the number
of points exceeds K, select a probability proportional to size sample
of points, output every point with an additional (non-descriptive)
`sampling.cardinality_limit.adjusted_count` attribute.
</details>

## Explanation

Consider a hypothetical telemetry signal in which an API event
produces a unit of data that has one or more associated numbers.
Using the OpenTelemetry Metrics data model terminology, we have two
scenarios in which sampling is common.

1. _Counter events:_ Each event represents a count, signifying the change in a sum.
2. _Histogram events:_ Each event represents an individual variable, signifying membership in a distribution.

> A Tracing Span event qualifies as both of these cases simultaneously.

Reviewer 8: This is awesome! 💯

Author: 😀

in the whole population, meaning all the events, using only the events
that were selected in the sample.

### Model and terminology

This model is meant to apply in telemetry collection situations where
individual events at an API boundary are sampled for collection.  Once
the process of sampling individual API-level events is understood, we
will learn to apply these techniques for sampling aggregated data.

In sampling, the term _sampling design_ refers to how sampling
probability is decided and the term _sample frame_ refers to how
events are organized into discrete populations.

For example, a simple design uses uniform probability, and a simple
framing technique is to collect one sample per distinct span name.

After executing a sampling design over a frame, each item selected in
the sample will have known _inclusion probability_, that determines
how likely the item was to being selected.  Implicitly, all the items
that were not selected for the sample have zero inclusion probability.

Descriptive words that are often used to describe sampling designs:

- *Fixed*: the sampling design is the same from one frame to the next

> - *Adaptive*: the sampling design changes from one frame to the next

Reviewer 8: I understand that these are hard to define.
Just curious if we want to consider something like "Adaptive: the sampling design changes from one frame to the next based on the input values". And here goes the reason:

If the sampling design changes dynamically based on the configuration from the operator, it sounds like it is "not fixed" but also "not adaptive".

Author: I added "based on the observed data".

The use of unbiased sampling outlined above makes it possible to
estimate the population total for arbitrary subsets of the sample, as
every individual sample has been independently assigned an adjusted
count.

There is a natural relationship between statistical bias and variance.
Approximate counting comes with variance, a matter of fact which can
be controlled for by the sample size.  Variance is unavoidable in an
unbiased sample, but it vanishes when you have enough data.

Although this makes it sound like small sample sizes are a problem,
due to expected high variance, this is just a limitation of the
technique.  When variance is high, use a larger sample size.

An easy approach for lowering variance is to aggregate sample frames
together across time.  For example, although the estimates drawn from
a one-minute sample may have high variance, combining an hour of
one-minute sample frames into an aggregate data set is guaranteed to
lower variance.  It must, because the data remains unbiased.

### Conveying the sampling probability

Some possibilities for encoding the adjusted count or inclusion
probability are discussed below, depending on the circumstances and

> the protocol.

Reviewer 8: Curious, is the protocol here specific to the exporting protocol, or it covers the RPC protocol where service A propagates the sampling decision+design to service B, so service B could provide consistent sampling decision instead of making its individual sampling decision?

It looks like we're only talking about the exporting protocol here.

Author: I added ```Here, the focus is on how to count sampled telemetry
+events in general, not a specific kind of event.  As we shall see in
+the following section, tracing comes with its addional complications.```

integer number in the range [0, +Inf).  This is a conceptually easy
way to understand sampling because larger numbers mean greater
representivity.

#### Encoding inclusion probability

We can encode the inclusion probability directly as a floating point
number in the range [0, 1).  This is typical of the Statsd format,
where each line includes an optional probability.  In this context,
the probability is also commonly referred to as a "sampling rate".  In
this case, smaller numbers mean greater representivity.

#### Encoding negative base-2 logarithm of inclusion probability

We can encode the negative base-2 logarithm of inclusion probability.
This restricts inclusion probabilities to powers of two and allows the
use of small non-negative integers to encode power-of-two adjusted
counts.  In this case, larger numbers mean exponentially greater
representivity.

#### Multiply the adjusted count into the data

When the data itself carries counts, such as for the Metrics Sum and
Histogram points.

This technique is less desirable because while it preserves the
expected value of the count or sum, the data loses information about

> variance.  This may also lead to rounding errors, when adjusted counts

Reviewer 8: +1, all the folks that I know + who made such choice, regret for it. This is especially painful when the metrics are used for alerting, the information loss about variance makes it hard to tell the confidence level of the alert.

Author: :+1:

- Produces complete traces
- Spans are countable.

#### Head sampling for traces

<details> <summary> Details about Sampler implementations that meet
the requirements stated above.  </summary>

##### `Parent` Sampler

The `Parent` Sampler ensures complete traces, provided all spans are
successfully recorded.  A downside of `Parent` sampling is that it
takes away control over Tracer overhead from non-roots in the trace.
To support counting spans, this Sampler requires propagating the
effective adjusted count of the context to use when starting child
spans.

In other head trace sampling schemes, we will see that it is useful to
propagate inclusion probability even for negative sampling decisions
(where the adjusted count is zero), therefore we prefer to use the
inclusion probability and not the adjusted count when propagating the
sampling rate via trace context.  The inclusion probability of a
context is referred to as its `head inclusion probability` for this
reason.

In addition to propagating head inclusion probability, to count
Parent-sampled spans, each span must directly encode its adjusted
count in the corresponding `SpanData`.  This may use a non-descriptive

> Resource or Span attribute named `sampling.parent.adjusted_count`, for

Reviewer: ```suggestion
Span attribute named `sampling.parent.adjusted_count`, for
```
Resource is supposed to be immutable

spans in a trace without using external information.  One approach,
[lost in the transition from OpenCensus to OpenTelemetry is to count
the number of children of each
span](https://github.com/open-telemetry/opentelemetry-specification/issues/355).

Lacking the number of expected children, we require a way to know the
minimum Sampler probability across traces to ensure they are complete.

To count TraceIDRatio-sampled spans, each span must encode its
adjusted count in the corresponding `SpanData`.  This may use a
non-descriptive Resource or Span attribute named
`sampling.traceidratio.adjusted_count`, for example.

##### Dapper's "Inflationary" Sampler

Google's [Dapper](https://research.google/pubs/pub36356/) tracing
system describes the use of sampling to control the cost of trace
collection at scale.  Dapper's early Sampler algorithm, referred to as
an "inflationary" approach (although not published in the paper), is
reproduced here.

This kind of Sampler allows non-root spans in a trace to raise the
probability of tracing, using a conditional probability formula shown
below.  Traces produced in this way are complete sub-trees, not

> necessarily complete.  This technique is succesful especially in

Reviewer 3: ```suggestion
necessarily complete.  This technique is successful especially in
```

The interpretation of "size", therefore, depends on what is being
estimated.  When sampling events with a natural size, such as for
Metric Sum data points, the absolute value of the point should be
multiplied with the adjusted count to form an effective input weight
(i.e., its "size" or contribution to the population total).  The
output adjusted count in this case is the output from weighted
sampling divided by the (absolute) point value.

#### Zero adjusted count

An adjusted count with zero value carries meaningful information,
specifically that the item participated in a probabilistic sampling
scheme and was not selected.  A zero value can be be useful to record
events when they provide useful information despite their effective
count; we can use this to combine multiple sampling schemes in a
single stream.

For example, consider collecting two samples from a stream of spans,
the first sample from those spans with attribute `A=1` and the second
sample from those spans with attribute `B=2`.  Any span that has both
of these properties is eligible to be selected for both samples, in
which case one event could have two non-zero adjusted counts (e.g.,

> `sampling.by_a.adjusted_count` and `sampling.by_b.adjusted_count`).

Reviewer: Another situation that gives rise to multiple weights is when different policies compete for the same trace. Eg the root service may have multiple sampling policies attached that have different purposes, with different sampling probabilities (and potentially different verbosity of perf data collection). When more than one policy "fires" they "share the body" of a single trace.

Anyway, it's not clear what is suggested here. If two weights are recorded, what happens to the suggestion of propagating the weight with the context downstream?

## Proposed Tracing specification

For the standard OpenTelemetry Span Sampler implementations to support
a range of probability sampling schemes, this document recommends the
use of a Span attribute named `sampling.sampler_name.adjusted_count`

> to encode an unbiased adjusted count computed by a Sampler.

Reviewer: What about situations where sampling is either biased or is subject to after-sampling rate limiting?

The value any attribute name prefixed with `sampling.` and suffixed
with `.adjusted_count` under this proposal MUST be an unbiased
estimate of the total population count represented by the individual
event.

### Suggested text

After
[Sampler](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#sampler)
and before [Builtin
Samplers](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#built-in-samplers),
a new section will introduce the term "adjusted count" and relate it
to inclusion probability.  For example:

```markdown
### Probability sampling

Probability Samplers are Samplers that output statistically unbiased
inclusion probability.  Inclusion probability in the context of

> tracing is the *effective* probability of the Sampler returning

Reviewer: The meaning of "Effective" is unclear

derived from the context when the [W3C Trace Context is-sampled
flag](https://www.w3.org/TR/trace-context/#sampled-flag) is in use.

#### Adjusted count span attribute

The recommended way to convey inclusion probability *for events*
sampled in OpenTelemetry is in the form of an **adjusted count**,
which is the reciprocal (i.e., mathematical inverse function) of
inclusion probability.

The implied goal of probability sampling is to support estimating the
count of spans in the population using the spans that were sampled.
The adjusted count associated with a Span is the expected value of the
number of identical Spans within the population that each Span
represents.  Probability samplers SHOULD ensure the samples they
compute are unbiased, which implies that the expected value of the sum
of adjusted counts in the sample equals the true count of spans in the
population.

Attributes used to express the adjusted count in an unbiased
probability sampling scheme SHOULD use a Span attribute name with
prefix `sampling.` and with suffix `.adjusted_count` (e.g.,
`sampling.sampler_name.adjusted_count`).  Adjusted count attributes
MAY be integer or floating-point values.

> #### Inclusion probability tracestate value

Reviewer: I think this should be optional

Reviewer 10: Can you clarify @yurishkuro ? It seems essential to me to convey the probability in _some_ form when exporting   to consumers of span data (like the collector  and it's various exporters)? Otherwise how can those exporters know how to upscale any statistics or metrics?

Reviewer: "essential to me" is the key part. This whole process (such as generating metrics on the fly from spans) may not be essential to some users, so the behavior should be opt-in.

Reviewer 10: > "some". 

I think _a majority_ of users would find this useful. It's not _just_ on the fly metric generation. Including the sampling rate _somewhere_ in the exported payload means providing users a more complete picture of the sampling decision that occurred at the time in the service/application/etc. Not doing that means we're giving them an incomplete picture. Doing on the fly metric generation is just one way that can then be used down the line. 

Could you give an example of a user who would prefer the tradeoff off off-by-default incomplete/missing sampling information for the benefit of not having to include/compute the extra key/value pair in their span data? I'm genuinely not trying to be snarky here I just am not sure I understand the "user story", so to speak. The use case I can think of (web-scale/perf costs/reduce payload size/etc) feels better served by `on-by-default, optional opt-out`.

Obviously I don't have the depth of knowledge you do here, but I think of sampling as a 1st class citizen in _tracing_ and what separates it from just structured logging, and so we should aim for as rich a sampling picture as is reasonable.

Anyway, hope that helps clarify what "essential to me" means.

Reviewer: Jaeger, Zipkin, OpenCensus SDKs do not propagate sampling weight. W3C Trace Context does not include the weight as a required part of the context. I think this provides enough evidence that this feature is not a requirement for a functioning tracing system.

I am completely on board with capturing the weight as a span attribute - all Jaeger SDKs do that, and any post-collection processing that hopes to build any aggregate data from traces would need that. But in-band propagation is not required for most use cases.

Reviewer 10: yup, per usual i'm an idiot and misread the previous section. plain ole span attribute works just fine for me, in-band (traceState) far less important and i don't have strong opinions here. txs for the patience.

Author: Having laid out the argument for propagating head inclusion probability, I can agree we should all be on the fence about it.

If the Parent sampler will be used and head inclusion probability is not propagated when the is-sampled bit is set, we will have a difficult time computing metrics from spans. However, the probability is only needed when is-sampled is true, so by turning on sampling the cost of propagating the sampling rate is automatically lowered.

Only the Dapper "inflationary" approach requires propagating the inclusion probability for negative sampling decisions, and that approach is not one I'm trying to defend. Unfortunately, we are left with few options that meet the three requirements stated in this document, which are:

1. Reduces Tracer overhead
2. Produces complete traces
3. Spans are countable.

OpenCensus was trying to switch to the use of TraceIDRatio sampling by default, which satisfies 1 and 3 easily. OpenTelemetry has a problem with 2, but I'm not sure it has to be solved. Completeness is less important than spans-to-metrics support, from our perspective as a vendor.

Author: I replaced the MUST with MAY to signify this is optional support.

The recommended way to convey sampling probability *for contexts* in
OpenTelemetry is through the W3C Trace Context tracestate using the

> key `head_probability`.

Reviewer: I think this is a common misunderstanding of tracestate header: it is not meant for arbitrary keys that have shared semantics. It is meant for vendor data.

What would be a valid use of tracestate is something like otel=headprob:0.123

Reviewer: I would be reluctant to bump the version of traceparent just to capture this value, since this can be done without changing w3c spec via `tracestate: otel=headprob:0.123`.

Author: I see, thanks. :+1: Can we say "using the key `otel` and sub-key `headprob`" or something similar?

Reviewer: > Can we say "using the key otel and sub-key headprob" or something similar?

yes

probability using tracestate, for the context that was in effect when
the W3C is-sampled bit was set.  The tracestate field SHOULD be set in
both sampled and unsampled cases, to convey the inclusion probability
even for unsampled contexts.

The `head_probability` tracestate key is set to a floating-point
number greater than or equal to 0 and less than or equal to 1.  The
floating point precision of the number SHOULD follow implementation
language standards and SHOULD be high enough to identify when Samplers
have different inclusion probabilities.

#### Counting probabilistically sampled spans

Consumers of a stream of span data that may or may not have been
sampled can follow these steps to count or approximately count the
total number of spans in the population.

For each span processed, locate all span attributes with prefix
`sampling.` and suffix `.adjusted_count`.  If there are no adjusted
count attributes on the span, count a single event.  If there is
exactly one adjusted count attribute, count that many identical span
events.

If there are more than one adjusted count attribute on the span, the
processor SHOULD make a consistent choice for spans of a given

> resource.  By default, the SDK's first preference MUST the builtin

Reviewer: a) why this specific priority?
b) this section describes downstream processor, why the reference to SDK?

Author: I threw this paragraph away and have removed all references to multiple adjusted counts.

I was trying to solve a problem when more than one framing is being used, but I see this was a mistake.

The key property is that if we expect to count all spans, we want to ensure the framing refers to all spans, so I've written that `sampling.adjusted_count` SHOULD refer to the whole population of spans.

Reviewer: still have the same two questions

Author: I meant to remove this section, it's now removed.

`TraceIDRatio` sampler, and its second preference MUST be the builtin
`Parent` sampler.
```

For the `TraceIDRatio` sampler, include the following additional text:

```md
When returning a `RECORD_AND_SAMPLE` decision, the TraceIDRatio

> Sampler MUST include the attribute

Reviewer 8: Would this be a breaking change for the spec (e.g. we will have to bump the spec version to 2.x)?

Reviewer 11: If it's because of the addition of `sampling.traceidratio.adjusted_count` attribute, I'd say we are fine as that's more related to semantic conventions (and we already 'handle' that part).

`RECORD_AND_SAMPLE` when invoked, which can be decided locally or 
derived from the context when the [W3C Trace Context is-sampled
flag](https://www.w3.org/TR/trace-context/#sampled-flag) is in use.

#### Adjusted count span attribute

The recommended way to convey inclusion probability *for events*
sampled in OpenTelemetry is in the form of an **adjusted count**,
which is the reciprocal (i.e., mathematical inverse function) of
inclusion probability.

The implied goal of probability sampling is to support estimating the
count of spans in the population using the spans that were sampled.
The adjusted count associated with a Span is the expected value of the
number of identical Spans within the population that each Span
represents.  Probability samplers SHOULD ensure the samples they
compute are unbiased, which implies that the sum
of adjusted counts in the sample approximates the true count of spans in the
population.

The adjusted count in an unbiased probability sampling scheme SHOULD
be expressed using a Span attribute named `sampling.adjusted_count`
when it represents the whole population of events.  Adjusted count 
attributes MAY be integer or floating-point values.

> #### Inclusion probability tracestate value

Reviewer: I am not sure it is clear how the SDKs should behave given that the propagation is optional. Propagating probability is a clear additional overhead that some sites may not want to pay, especially if they don't use any inflationary or partial sampling. So how does an SDK know if it should or should not include the probability in the tracestate? Should it be a configuration of the SDK (tricky because guaranteeing identical configuration throughout all microservices is quite difficult)?

Author: The text is outdated, resolving. I am now proposing non-optional support. See https://github.com/open-telemetry/oteps/pull/168

The recommended way to convey sampling probability *for contexts* in
OpenTelemetry is through the W3C Trace Context tracestate using the
key `head_probability`.

Probability Samplers MAY encode the effective sampling inclusion
probability using tracestate, for the context that was in effect when

> the W3C is-sampled bit was set.  The tracestate field SHOULD be set in

Reviewer: > The tracestate field SHOULD be set in

This needs to be qualified, so that SHOULD is interpreted conditionally, e.g.

> **If the inclusion probability is propagated**, it SHOULD be represented with subkey `headprob` within the field `otel`... For example:
```
tracestate: otel=headprob:0.1234
```

both sampled and unsampled cases, to convey the inclusion probability
even for unsampled contexts.

> The `head_probability` tracestate key is set to a floating-point

Reviewer: ```suggestion
The `headprob` subkey is set to a floating-point
```

number greater than or equal to 0 and less than or equal to 1.  The
floating point precision of the number SHOULD follow implementation
language standards and SHOULD be high enough to identify when Samplers
have different inclusion probabilities.

#### Counting probabilistically sampled spans

Consumers of a stream of span data that may or may not have been
sampled can follow these steps to count or approximately count the
total number of spans in the population.

For each span processed, locate the `sampling.adjusted_count` attribute. 
If there is none, count a single span event.  If the attribute is set, 
count that many identical span events.
```

For the `TraceIDRatio` sampler, include the following additional text:

```md
When returning a `RECORD_AND_SAMPLE` decision, the TraceIDRatio
Sampler MUST include the attribute `sampling.adjusted_count=C`, where
`C` is the reciprocal of the configured trace ID ratio.

The returned tracestate used for the child context MAY have the
tracestate `otel` key with the sub-key `headprob` set to the configured 
trace ID ratio.
```

For the `Parent` sampler, include the following additional text:

```md

> When returning a `RECORD_AND_SAMPLE` decision, the Parent Sampler MAY

Reviewer: a few things here:
  1. parent-present and parent-absent must be treated differently, the attribute should only be set when parent is present, otherwise defer to the fallback sampler
  2. MAY keyword for recording the attribute seems unnecessary weak, considering that ratio sampler has MUST for the same
  3. on the other hand, the parent probability propagation is optional, so it could be MUST but with a qualification

Author: Thanks for your help. I tossed most of this text and came up with https://github.com/open-telemetry/oteps/pull/168

include the attribute `sampling.adjusted_count=C`, where `C` is the
reciprocal of the parent trace context's head inclusion probability.

The tracestate `otel` key with the sub-key `headprob` is used to lookup
and propagate the configured  trace ID ratio.  When the `otel` key with 

> sub-key `headprob` is not located and `is-sampled` is set, the Sampler 

Reviewer: the tracestate seems to be described multiple times, why not do it once only?

Author: I realized that my employer is definitely not interested in optional support, so I have returned with a mandatory but limited proposal here: https://github.com/open-telemetry/oteps/pull/168

> MUST set `sampling.adjusted_count=0` to signal that spans cannot be 

Reviewer: How would setting it to 0 be different from not setting at all?

If we look at the existing trace context formats like Zipkin's B3 or Jaeger's, they do not propagate probability and they use classic head-based sampling with sampling decision propagated. Extrapolation from individual spans is not possible with those, since there's no way of knowing probability or count of children spans. Extrapolation from re-assembled trace is possible as long as the count is captured in the root. I think whichever behavior we propose, it should degrade nicely into this legacy (and very widespread) mode.

Author: The text is outdated, resolving.

a new section will introduce the term "adjusted count" and relate it
to inclusion probability.  For example:

```markdown
### Probability sampling

Probability Samplers are Samplers that output statistically unbiased
inclusion probability.  Inclusion probability in the context of
tracing is the *effective* probability of the Sampler returning
`RECORD_AND_SAMPLE` when invoked, which can be decided locally or 
derived from the context when the [W3C Trace Context is-sampled
flag](https://www.w3.org/TR/trace-context/#sampled-flag) is in use.

#### Adjusted count span attribute

The recommended way to convey inclusion probability *for events*
sampled in OpenTelemetry is in the form of an **adjusted count**,
which is the reciprocal (i.e., mathematical inverse function) of
inclusion probability.

The implied goal of probability sampling is to support estimating the
count of spans in the population using the spans that were sampled.
The adjusted count associated with a Span is the expected value of the
number of identical Spans within the population that each Span
represents.  Probability samplers SHOULD ensure the samples they

> compute are unbiased, which implies that the expected value of the sum
> of adjusted counts in the sample equals the true count of spans in the

Reviewer: ```suggestion
compute are unbiased, which implies that the sum
of adjusted counts in the sample approximates the true count of spans in the
```

population.

The adjusted count in an unbiased probability sampling scheme SHOULD
be expressed using a Span attribute named `sampling.adjusted_count`
when it represents the whole population of events.  Adjusted count 
attributes MAY be integer or floating-point values.

#### Inclusion probability tracestate value

The recommended way to convey sampling probability *for contexts* in
OpenTelemetry is through the W3C Trace Context tracestate using the
key `head_probability`.

> Probability Samplers SHOULD encode the effective sampling inclusion

Reviewer: ```suggestion
Probability Samplers MAY encode the effective sampling inclusion
```

probability using tracestate, for the context that was in effect when
the W3C is-sampled bit was set.  The tracestate field SHOULD be set in
both sampled and unsampled cases, to convey the inclusion probability
even for unsampled contexts.

The `head_probability` tracestate key is set to a floating-point
number greater than or equal to 0 and less than or equal to 1.  The
floating point precision of the number SHOULD follow implementation
language standards and SHOULD be high enough to identify when Samplers
have different inclusion probabilities.

#### Counting probabilistically sampled spans

Consumers of a stream of span data that may or may not have been
sampled can follow these steps to count or approximately count the
total number of spans in the population.

For each span processed, locate the `sampling.adjusted_count` attribute. 

> If there is none, count a single span event.  If the attribute is set, 

Reviewer: > If there is none, count a single span event.

I don't think this is correct, because spans inherit adjusted count from their parent

Author: There are several ways to interpret a count. We can count spans by the adjusted count of all the trace roots, or we can count spans by the adjusted count of all the spans independently. I aimed to talk about counting spans in their own right in this section, without referring to trace roots.

Spans may be counted using the adjusted count of the parent, in which case once in a while the parent's adjusted count comes through along with `is-sampled`. This also implies that much of the time `is-sampled` is not set and a zero adjusted count is assumed.

Spans may be counted by sampling at the call site independently, in which case if there is no sampling being applied, there will be no adjusted count set and the correct count is 1.

You are correct, I have failed to call out an error scenario, when `is-sampled` is set and the `otel=headprob:` value is unknown. I've written that the Parent Sampler MUST set `adjusted_count=0`, to signal that they must be counted using trace roots, not using span adjusted counts.

That way, when there is no sampling taking place, 100% of spans will be recorded and counted as one span each.

count.

When the interval expires and the sample frame is considered complete,
the selected sample spans are output with possibly updated adjusted
counts.
</details>

### Metric sampling

<details> 
<summary> Example use-cases for probability sampling of metrics
are aimed at lowering cost and addressing high cardinality.</summary>

#### Statsd Counter

A Statsd counter event appears as a line of text, describing a
number-valued event with optional attributes and inclusion probability
("sample rate").

For example, a metric named `name` is incremented by `increment` using
a counter event (`c`) with the given `sample_rate`.

```
name:increment|c|@sample_rate
```

For example, a count of 100 that was selected for a 1-in-10 simple
random sampling scheme will arrive as:

```
counter:100|c|@0.1
```

Events in the example have with 0.1 inclusion probability have
adjusted count of 10.  Assuming the sample was selected using an
unbiased algorithm, we can interpret this event as having an expected
count of `100/0.1 = 1000`.

#### Metric exemplars with adjusted counts

The OTLP protocol for metrics includes a repeated exemplars field in

> every data point.  This is a place where histogram implementations are

Reviewer: Do exemplars only apply to histograms?

make it possible to approximately count events using dimensions that
were dropped during metric aggregation.

An end-to-end pipeline of sampled metrics events can be constructed
based on exemplars with adjusted counts, one capable of supporting
approximate-count queries over sampled metric events at high
cardinality.

#### Metric cardinality limiter

A metrics processor can be configured to limit cardinality for a
single metric name, allowing no more than K distinct label sets per
export interval.  The export interval is fixed to a short interval so
that a complete set of distinct labels can be stored temporarily.

Caveats: as presented, this works for Sum and Histogram points
received with Delta aggregation temporality and where the Sum is
monotonic (see
[opentelemetry-proto/issues/303](https://github.com/open-telemetry/opentelemetry-proto/issues/303)).

Considering data points received during the interval, when the number
of points exceeds K, select a probability proportional to size sample
of points, output every point with an additional (non-descriptive)
`sampling.cardinality_limit.adjusted_count` attribute.
</details>

## Explanation

> Consider a hypothetical telemetry signal in which an API event

Reviewer: What is an API event?

2. _Histogram events:_ Each event represents an individual variable, signifying membership in a distribution.

A Tracing Span event qualifies as both of these cases simultaneously.
One span can be interpreted as at least one Counter event (e.g., one
request, the number of bytes read) and at least one Histogram event
(e.g., request latency, request size).

In Metrics, [Statsd Counter and Histogram events meet this definition](https://github.com/statsd/statsd/blob/master/docs/metric_types.md#sampling).

In both cases, the goal in sampling is to estimate the count of events
in the whole population, meaning all the events, using only the events
that were selected in the sample.

### Model and terminology

This model is meant to apply in telemetry collection situations where
individual events at an API boundary are sampled for collection.  Once
the process of sampling individual API-level events is understood, we
will learn to apply these techniques for sampling aggregated data.

In sampling, the term _sampling design_ refers to how sampling
probability is decided and the term _sample frame_ refers to how

> events are organized into discrete populations.

Reviewer: Unclear what sample frame is. Is it a time interval? That's how it seems to be used in examples (eg "over a frame"), except for the next paragraph.

inclusion count for N each, approximately speaking.

This intuition is backed up with statistics.  This equation is known
as the Horvitz-Thompson estimator of the population total, a
general-purpose statistical "estimator" that applies to all _without
replacement_ sampling designs.

Assuming sample data is correctly computed, the consumer of sample
data can treat every sample event as though an identical copy of
itself has occurred _adjusted count_ times.  Every sample event is
representative for adjusted count many copies of itself.

There is one essential requirement for this to work.  The selection
procedure must be _statistically unbiased_, a term meaning that the
process is required to give equal consideration to all possible
outcomes.

#### Sampling and variance

The use of unbiased sampling outlined above makes it possible to
estimate the population total for arbitrary subsets of the sample, as
every individual sample has been independently assigned an adjusted
count.

There is a natural relationship between statistical bias and variance.
Approximate counting comes with variance, a matter of fact which can
be controlled for by the sample size.  Variance is unavoidable in an

> unbiased sample, but it vanishes when you have enough data.

Reviewer: This sounds like there's some magic threshold for sample size. Maybe rephrase that variance diminishes with growing sample size.

Although this makes it sound like small sample sizes are a problem,
due to expected high variance, this is just a limitation of the
technique.  When variance is high, use a larger sample size.

> An easy approach for lowering variance is to aggregate sample frames

Reviewer: Not following this. The way to decrease variance is to increase the ratio of sample size vs population size. Aggregating 1m to 1hr does not change that ratio.

counts.  In this case, larger numbers mean exponentially greater
representivity.

#### Multiply the adjusted count into the data

When the data itself carries counts, such as for the Metrics Sum and
Histogram points.

This technique is less desirable because while it preserves the
expected value of the count or sum, the data loses information about
variance.  This may also lead to rounding errors, when adjusted counts
are not integer valued.

### Trace Sampling

Sampling techniques are always about lowering the cost of data
collection and analysis, but in trace collection and analysis
specifically, approaches can be categorized by whether they reduce
Tracer overhead.  Tracer overhead is reduced by not recording spans
for unsampled traces and requires making the sampling decision for a
trace before all of its attributes are known.

Traces are expected to be complete, meaning that a tree or sub-tree of
spans branching from a certain root are expected to be fully
collected.  When sampling is applied to reduce Tracer overhead, there
is generally an expectation that complete traces will still be
produced.  Sampling techniques that lower Tracer overhead and produce

> complete traces are known as _Head trace sampling_ techniques.

Reviewer: ```suggestion
complete traces are known as _Head-based trace sampling_ techniques.
```

The decision to produce and collect a sample trace has to be made when
the root span starts, to avoid incomplete traces.  Then, assuming
complete traces can be collected, the adjusted count of the root span
determines an adjusted count for every span in the trace.

#### Counting child spans using root span adjusted counts

The adjusted count of a root span determines the adjusted count of
each of its children based on the following logic:

> - The root span is considered representative of _adjusted count_ many

Reviewer: The phrasing tripped me a bit, I would use adjusted_count (fixed width font) to indicate that we're referring to a single variable, which space separated italic does not convey.

(i.e., its "size" or contribution to the population total).  The
output adjusted count in this case is the output from weighted
sampling divided by the (absolute) point value.

#### Zero adjusted count

An adjusted count with zero value carries meaningful information,
specifically that the item participated in a probabilistic sampling
scheme and was not selected.  A zero value can be be useful to record
events when they provide useful information despite their effective
count; we can use this to combine multiple sampling schemes in a
single stream.

For example, consider collecting two samples from a stream of spans,
the first sample from those spans with attribute `A=1` and the second
sample from those spans with attribute `B=2`.  Any span that has both
of these properties is eligible to be selected for both samples, in
which case one event could have two non-zero adjusted counts (e.g.,
`sampling.by_a.adjusted_count` and `sampling.by_b.adjusted_count`).

## Proposed Tracing specification

For the standard OpenTelemetry Span Sampler implementations to support
a range of probability sampling schemes, this document recommends the

> use of a Span attribute named `sampling.sampler_name.adjusted_count`

Reviewer: I would prefer a different naming scheme. Trace processing pipeline does not know which sampler is used for each individual span, and in most cases it doesn't care either as long as the adjusted count is accurate. But since the attribute name includes the sampler name, the processor has no way of retrieving the count via O(1) lookup, it must iterate through all attributes with prefix match.
